src/govuk/components/header/header.mjs

```diff
@@ -8,34 +8,63 @@ function Header ($module) {
   this.$menu = this.$menuButton && $module.querySelector(
     '#' + this.$menuButton.getAttribute('aria-controls')
   )
+
+  // Save the opened/closed state for the nav in memory so that we can
+  // accurately maintain state when the screen is changed from small to
+  // big and back to small
+  this.menuIsOpen = false
+
+  // A global const for storing a matchMedia instance which we'll use to
+  // detect when a screen size change happens. We set this later during the
+  // init function and rely on it being null if the feature isn't available
+  // to initially apply hidden attributes
+  this.mql = null
 }
 
 /**
  * Initialise header
  *
  * Check for the presence of the header, menu and menu button – if any are
  * missing then there's nothing to do so return early.
+ * Additionally feature sniff for and apply a matchMedia for desktop which will
+ * trigger a state sync if the browser viewport moves between states
  */
 Header.prototype.init = function () {
   if (!this.$module || !this.$menuButton || !this.$menu) {
     return
   }
 
-  this.syncState(this.$menu.classList.contains('govuk-header__navigation-list--open'))
+  if ('matchMedia' in window) {
+    this.mql = window.matchMedia('(min-width: 48.0625em)')
+    this.mql.addEventListener('change', this.syncState.bind(this))
+  }
+
+  this.syncState()
   this.$menuButton.addEventListener('click', this.handleMenuButtonClick.bind(this))
 }
 
 /**
  * Sync menu state
  *
- * Sync the menu button class and the accessible state of the menu and the menu
- * button with the visible state of the menu
- *
- * @param {boolean} isVisible Whether the menu is currently visible
+ * Uses the global variable menuIsOpen to correctly set the accessible and
+ * visual states of the menu and the menu button.
+ * Additionally will force the menu to be visible and the menu button to be
+ * hidden if the match media is triggered
  */
-Header.prototype.syncState = function (isVisible) {
-  this.$menuButton.classList.toggle('govuk-header__menu-button--open', isVisible)
-  this.$menuButton.setAttribute('aria-expanded', isVisible)
+Header.prototype.syncState = function () {
+  if (this.mql === null || !this.mql.matches) {
+    this.$menuButton.removeAttribute('hidden')
+    this.$menuButton.setAttribute('aria-expanded', this.menuIsOpen)
+
+    if (this.menuIsOpen) {
+      this.$menu.removeAttribute('hidden')
+    } else {
+      this.$menu.setAttribute('hidden', '')
+    }
+  } else if (this.mql === null || this.mql.matches) {
+    this.$menu.removeAttribute('hidden')
+    this.$menuButton.setAttribute('hidden', '')
+  }
```

Comment:
I find it a little hard to follow what's going on with the logic here, and I'm not sure it's doing what you want.

- `this.mql === null` is included in both the `if` and `if else` part of the conditional, and so can never be true in the `else if` (because it `this.mql === null` we'd always go into the `if` part of the conditional and never get to the `else`)
- `this.mql.match` can also only be `true` or `false`, so there's no need for the `else if` – it should just be `else`.

It might be easier to make this understandable if we either flip the logic or update the media query to use `max-width` so it matches on mobile/tablet.

What is it that we actually want to happen when the browser doesn't support `matchMedia`?

Reply:
The idea is that if `matchMedia` isn't supported it will only ever run the contents of that first `if` block which manages changes in state for mobile view. If it is supported and `syncState` is triggered via the `mql` `change` event, then it'll run the contents of the `else` which essentially clear the UI, hiding the mobile menu button and forcing the menu to be visible.

This has prompted me to check this on an older browser which don't support `matchMedia` and the menu isn't visible because it never reaches that attribute clearing stage 😬 Also when I look at this again I agree that there's no reason for that `else if` to be there. I'm gonna clear this up and switch it as suggested.

It's also probably worth making this change on the design system website where I lifted this solution from which I can do separately.

Reply:
I've figured it out. The `this.mql === null` was a way to let the user use the menu and let the syncing function work without having `matchMedia` enabled and without having to repeat code for if the user had `matchMedia` or not since `this.mql.matches` errors if `mql` is null. There's obviously a tidier, less confusing way to do this.

Reply:
This ended up being a bit of a can of worms. In trying to account for the user not having `matchMedia` available, we create a risk that users of those browsers wont be able to use the header navigation. For example if a user is on IE9 (doesn't support `matchMedia`) and JS does load but CSS doesn't, then the JS will add `hidden` to the menu, as we don't check screen size by any means other than match media so assume they're on mobile to let them use the mobile menu, and without the CSS overrides to force the menu to display, the menu will be hidden for no CSS users.

I can see 2 ways around this whilst still solving the linked issue:
1. Present a "no JS" view to users who don't have `matchMedia` available
2. Don't user `matchMedia`

I've opted for 1 as the number of IE9 and below users are very low, a pertinent point as we're planning to reduce support for IE11 and below in v5, they are unlikely to be on screens small enough to require the mobile menu UI and they can still access the links in the menu this way.